### PR TITLE
Add function to show avatar image for messages

### PR DESCRIPTION
**Pull request by zwang**
Demo app updated to show avatar image only for text messages by modifying the viewModelBuilder, Presenter is probably best place to set avatar image based on viewmodel and decoration attributes, but this is highly case by case. 

By default, the avatar is at the bottom of the collection view cell. No way to customize it unless changing code.

Screenshot of demo app as below:
simulator screen shot mar 10 2016 12 08 20 am
simulator screen shot mar 10 2016 12 08 14 am
simulator screen shot mar 9 2016 11 50 02 pm

### CODE REVIEW

File: ChattoAdditions/Source/Chat Items/BaseMessage/Views/BaseMessageCollectionViewCell.swift [hunk 1]
```diff
@@ -194,6 +203,13 @@ public class BaseMessageCollectionViewCell<BubbleViewType where BubbleViewType:U
         self.bubbleView.preferredMaxLayoutWidth = layoutModel.preferredMaxWidthForBubble
         self.bubbleView.layoutIfNeeded()
 
+        let avatarImageSize = baseStyle.getAvatarImageSize(self.messageViewModel)
+        if avatarImageSize != CGSize.zero {
+            self.avatarView.bma_rect = layoutModel.avatarViewFrame
+            self.avatarView.image = self.messageViewModel.avatarImage
+            self.avatarView.layoutIfNeeded()
```
Comment (diegosanchezr) [inline ChattoAdditions/Source/Chat Items/BaseMessage/Views/BaseMessageCollectionViewCell.swift]: Does simply doing  self.avatarView.bma_rect = layoutModel.avatarViewFrame works fine? Can you move setting the image to updateViews?

File: ChattoAdditions/Source/Chat Items/BaseMessage/Views/BaseMessageCollectionViewCellDefaultStyle.swift [hunk 2]
```diff
@@ -32,23 +32,23 @@ public class BaseMessageCollectionViewCellDefaultStyle: BaseMessageCollectionVie
...
     }()
 
     lazy var borderOutgoingNoTail: UIImage = {
-        return UIImage(named: "bubble-outgoing-border", inBundle: NSBundle(forClass: self.dynamicType), compatibleWithTraitCollection: nil)!
+        return UIImage(named: "bubble-outgoing-border", inBundle: NSBundle(forClass: BaseMessageCollectionViewCellDefaultStyle.self), compatibleWithTraitCollection: nil)!
     }()
 
     public lazy var failedIcon: UIImage = {
-        return UIImage(named: "base-message-failed-icon", inBundle: NSBundle(forClass: self.dynamicType), compatibleWithTraitCollection: nil)!
+        return UIImage(named: "base-message-failed-icon", inBundle: NSBundle(forClass: BaseMessageCollectionViewCellDefaultStyle.self), compatibleWithTraitCollection: nil)!
```
Comment (diegosanchezr) [inline ChattoAdditions/Source/Chat Items/BaseMessage/Views/BaseMessageCollectionViewCellDefaultStyle.swift]: Mmm I see what happened here; we should probably make this change all over the framework 👍

File: ChattoApp/ChattoApp/Assets.xcassets/userAvatar.imageset/Contents.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "idiom" : "universal",
+      "scale" : "1x"
+    },
+    {
+      "idiom" : "universal",
+      "filename" : "userAvatar@2x.png",
```
Comment (diegosanchezr) [inline ChattoApp/ChattoApp/Assets.xcassets/userAvatar.imageset/Contents.json]: Any chance of having a vectorized pdf?

File: ChattoApp/ChattoApp/Source/BaseMessageCollectionViewCellAvatarStyle.swift [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import Foundation
+import ChattoAdditions
+
+class BaseMessageCollectionViewCellAvatarStyle: BaseMessageCollectionViewCellDefaultStyle {
+    override func getAvatarImageSize(messageViewModel: MessageViewModelProtocol) -> CGSize {
+        // Display avatar for both incoming and outgoing messages for demo purpose
+        if messageViewModel.isIncoming {
+            return CGSize(width: 35, height: 35)
+        } else {
+            return CGSize(width: 35, height: 35)
```
Comment (diegosanchezr) [inline ChattoApp/ChattoApp/Source/BaseMessageCollectionViewCellAvatarStyle.swift]: You may want to remove the if/else

File: ChattoApp/ChattoApp/Source/Text Messages/DemoTextMessageViewModel.swift [hunk 5]
```diff
@@ -44,6 +44,9 @@ public class DemoTextMessageViewModelBuilder: ViewModelBuilderProtocol {
     public func createViewModel(textMessage: DemoTextMessageModel) -> DemoTextMessageViewModel {
         let messageViewModel = self.messageViewModelBuilder.createMessageViewModel(textMessage)
         let textMessageViewModel = DemoTextMessageViewModel(textMessage: textMessage, messageViewModel: messageViewModel)
+        // Best place to decide whether user having avart might be in the presenter. Here just for demo purpose
+        // Because we might only want to display avatar when showTail in the nessage bubble
+        textMessageViewModel.avatarImage = UIImage(named: "userAvatar")
```
Comment (diegosanchezr) [inline ChattoApp/ChattoApp/Source/Text Messages/DemoTextMessageViewModel.swift]: Mmm... Haven't thought of this. You're right, trying to implement some kind of logic in which the avatar is not shown in every cell looks complicated as it is now. I'd rather add a new attribute ChatItemDecorationAttributes.canShowAvatar and replicate what it's done with the tail

**Comment by zwang** [comment]
@diegosanchezr thanks for the review and notes. I will revisit them this weekend and update the PR then. Thanks again :)

File: ChattoApp/ChattoApp/Source/Text Messages/DemoTextMessageViewModel.swift (hunk 5 above)
Comment (zwang) [inline ChattoApp/ChattoApp/Source/Text Messages/DemoTextMessageViewModel.swift]: I am not sure how adding `canShowAvatar` to the `ChatItemDecorationAttributes` can make things easier yet. I need to look into the code to see more use cases of `ChatItemDecorationAttributes`. So if it makes sense, I will add this to the decorator attributes later. :) 

I think user can also do this by themselves by subclassing `ChatItemDecorationAttributes`?

However for now, I will suggest to do this in the presenter based on the actual logic of view model and business requirements. User need to create an custom presenter.

**Comment by zwang** [comment]
@diegosanchezr thanks a lot for the review and feedback again. It all makes sense. I am hesitate to change based on the last note as I commented above. So I will leave it there for now and get more familiar with the code first. 

Other than the last note, I should have resolved all issues in all the other notes. And also I add two new small features: 1. Allow user to subclass the basestyle to specify `VerticalAlignment` for the avatar, like Top or Center, by default it is Bottom. 2. Allow user to subclass the basestyle to change the default `BaseMessageCollectionViewCellLayoutConstants`. These two features are both needed for my project. And I think they should be helpful to others as well. It remains simple and yet allow others to extend to customize the base collection view cell.

Thanks again for you help. Cheers! :)

**Comment by diegosanchezr** [comment]
@zwang thank you very much again! merging!

**Comment by zwang** [comment]
@diegosanchezr  Thank you. :)

**Comment by aytunch** [comment]
@zwang Hi and thanks for your effort in this feature. I have a couple questions. I downloaded the dev branch and the Avatars seems to be only next to text typed messages but not image typed ones. Is this a design choice or are you still working on adding them to all message types?

Since there is a `senderId` in `MessageModelProtocol`, can we display different avatars according to the `senderId`'s? Does that mean that the feature you are working on is not just Avatar support and also covers group chat support as @diegosanchezr pointed out in the issue I have opened named: "Group Chat Support"? If thats the case, then images will need the avatars to be next to them.

I really believe Group chat is a vital part of chat and it should be present in this beautiful Library:)

Last question will be about the avatar views. Are they just `UIImage`'s? or `UIView`s? the reason I am asking is if they are `UIView`'s, It would be easy to display a `UILabel` beneath the avatars which is necessary for group chat. I realize there is not much horizontal space for the names but at least initials can be displayed or we can use small fonts just to inform the user. I don't think just the avatars are enough, since it is a big possibility that people can have same avatars or no avatars set at all.

Thanks to all of you who are keeping this Library alive:)

**Comment by zwang** [comment]
Hi @aytunch 

First, the feature only supports adding avatar to messages. Not related to group chat. But should work in group chat too. It is just a matter of setting the avatar image for each message. 

With this said, it allow you to put avatar image next to photo messages too. I just didn't do it in the demo app. You can checkout `DemoTextMessageViewModel.swift` line 49 and do the same thing for photo message in `DemoPhotoMessageViewModel.swift` just for a quick demo.

You definitely can display different avatar image according to senderId. What I did in the demo app is just an example, you can set the `avatarImage` value of `MessageViewModelProtocol` in any place as long as it make sense to you. For myself, I did in each message presenter. 

Regarding the sender name under avatar, I think you can do the same extension as I did for avatar view. I would not recommend to have name bound together with avatar view. They can stay together, but they should not be bound together. So the avatar view is just an `UIImageView` holding images. An good idea might be to display the name right above the bubble. 

Hope my answer helps. :)

**Pull request merged**